File: apps/android/app/src/main/java/com/cursumi/app/core/model/AdminModels.kt
package com.cursumi.app.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import java.util.UUID

// Instructor y administración.

@Serializable data class InstructorEarnings(val total: Double = 0.0, val thisMonth: Double? = null, val courses: Int? = null)
@Serializable data class InstructorAnalytics(val totalCourses: Int = 0, val publishedCourses: Int = 0, val totalStudents: Int = 0, val avgProgress: Double = 0.0)

@Serializable
data class InstructorCourse(
    val id: String, val title: String, val modality: String? = null, val status: String = "draft",
    val price: Double? = null, val imageUrl: String? = null, val studentsCount: Int? = null,
) {
    val statusLabel get() = mapOf("draft" to "Borrador", "published" to "Publicado", "archived" to "Archivado")[status] ?: status
}

@Serializable data class ConvStudent(val name: String? = null)
@Serializable data class ConvCourse(val title: String? = null)
@Serializable data class ConvMessage(val body: String)
@Serializable data class InstructorConversation(val id: String, val student: ConvStudent? = null, val course: ConvCourse? = null, val messages: List<ConvMessage>? = null)

@Serializable data class InstructorProfile(val headline: String = "", val bio: String = "", val specialties: String = "", val teachingYears: Int? = null)
@Serializable data class InstructorProfileUpdate(val headline: String, val bio: String, val specialties: String, val teachingYears: Int? = null)
@Serializable data class StripeStatus(val connected: Boolean? = null, val onboarded: Boolean? = null)

@Serializable
data class NewLesson(
    val id: String = UUID.randomUUID().toString(), val title: String = "", val type: String = "text",
    val order: Int = 0, val content: String? = null, val videoUrl: String? = null,
)
@Serializable data class NewSection(val id: String = UUID.randomUUID().toString(), val title: String = "", val order: Int = 0, val lessons: List<NewLesson> = emptyList())

@Serializable
data class NewCoursePayload(
    val title: String, val description: String, val category: String, val level: String, val modality: String,
    /** Derivado de la modalidad: virtual → ondemand · evento → fechado */
    val courseType: String, val startDate: String, val duration: String, val price: Double,
    val imageUrl: String? = null, val sections: List<NewSection>, val isDraft: Boolean,
)

@Serializable data class HostGameCount(val participants: Int? = null, val questions: Int? = null)
@Serializable data class HostGame(val id: String, val title: String, val code: String = "", val status: String = "", @SerialName("_count") val count: HostGameCount? = null)

@Serializable
data class NewGameQuestion(
    @Transient val uid: String = UUID.randomUUID().toString(),
    val question: String = "", val options: List<String> = listOf("", "", "", ""), val correct: Int = 0,
)

// ── Admin ────────────────────────────────────────────────────────────────────

@Serializable
data class AdminStats(
    val totalUsers: Int = 0, val totalCourses: Int = 0, val publishedCourses: Int = 0, val draftCourses: Int = 0,
    val totalEnrollments: Int = 0,
    /** Valor de catálogo (precio × inscripciones), NO dinero cobrado. */
    val estimatedRevenue: Double = 0.0,
)
@Serializable data class AdminFinances(val totalRevenue: Double? = null, val totalPlatformFee: Double? = null, val totalInstructorPayouts: Double? = null, val thisMonthRevenue: Double? = null)
@Serializable data class RevenueMonth(val month: String, val amount: Double = 0.0)
@Serializable data class UsersMonth(val month: String, val users: Int = 0)
@Serializable data class AdminAnalytics(val revenueByMonth: List<RevenueMonth>? = null, val usersByMonth: List<UsersMonth>? = null)

@Serializable data class TitledCourse(val title: String? = null)
@Serializable data class AdminReview(val id: String, val rating: Int = 0, val comment: String? = null, val approved: Boolean = false, val user: NamedUser? = null, val course: TitledCourse? = null)
@Serializable data class AppUser(val id: String, val name: String? = null, val email: String? = null)
@Serializable data class AdminApplication(val id: String, val status: String = "", val headline: String? = null, val bio: String? = null, val reason: String? = null, val user: AppUser? = null)
@Serializable data class AdminUser(val id: String, val name: String? = null, val email: String? = null, val role: String = "student")
@Serializable data class AdminCoupon(val id: String, val code: String, val discountPct: Int = 0, val maxUses: Int? = null, val usedCount: Int? = null, val active: Boolean = true)
@Serializable data class CatCount(val courses: Int? = null)
@Serializable data class AdminCategory(val id: String, val name: String, val slug: String = "", @SerialName("_count") val count: CatCount? = null)
@Serializable
data class AdminKpi(val id: String, val name: String, val unit: String? = null, val targetValue: Double = 0.0, val currentValue: Double = 0.0) {
    val percent get() = if (targetValue > 0) minOf(100, Math.round(currentValue / targetValue * 100).toInt()) else 0
}

@Serializable
data class QuoteRequest(
    val id: String, val companyName: String, val contactName: String = "", val contactEmail: String = "",
    val contactPhone: String? = null, val companySize: String? = null, val interests: String? = null,
    val message: String? = null, val status: String = "new",
) {
    val statusLabel get() = mapOf("new" to "Nueva", "contacted" to "Contactada", "converted" to "Convertida", "closed" to "Cerrada")[status] ?: status
}

@Serializable
data class QuoteRequestPayload(
    val companyName: String, val contactName: String, val contactEmail: String, val contactPhone: String? = null,
    val companySize: String? = null, val interests: String? = null, val message: String? = null,
)

File: apps/ios/Cursumi/Core/Models/AdminModels.swift
import Foundation

// Instructor y administración. Espejo de `apps/mobile/src/lib/me.ts`.

/// Lista que la API devuelve como array suelto o envuelta en `{ <clave>: [...] }`.
struct FlexibleList<T: Decodable>: Decodable {
    let items: [T]

    init(from decoder: Decoder) throws {
        if let list = try? decoder.singleValueContainer().decode([T].self) {
            items = list
            return
        }
        let container = try decoder.container(keyedBy: AnyKey.self)
        for key in container.allKeys {
            if let list = try? container.decode([T].self, forKey: key) {
                items = list
                return
            }
        }
        items = []
    }

    private struct AnyKey: CodingKey {
        var stringValue: String
        var intValue: Int? { nil }
        init?(stringValue: String) { self.stringValue = stringValue }
        init?(intValue: Int) { nil }
    }
}

// MARK: - Instructor

struct InstructorEarnings: Decodable {
    let total: Double
    let thisMonth: Double?
    let courses: Int?
}

struct InstructorAnalytics: Decodable {
    let totalCourses: Int
    let publishedCourses: Int
    let totalStudents: Int
    let avgProgress: Double
}

struct InstructorCourse: Decodable, Identifiable, Equatable {
    let id: String
    let title: String
    let modality: String?
    let status: String // draft | published | archived
    let price: Double?
    let imageUrl: String?
    let studentsCount: Int?

    var statusLabel: String {
        ["draft": "Borrador", "published": "Publicado", "archived": "Archivado"][status] ?? status
    }
}

struct InstructorConversation: Decodable, Identifiable {
    struct Student: Decodable { let name: String? }
    struct Course: Decodable { let title: String? }
    struct Message: Decodable { let body: String }
    let id: String
    let student: Student?
    let course: Course?
    let messages: [Message]?
}

struct InstructorProfile: Decodable {
    let headline: String
    let bio: String
    let specialties: String
    let teachingYears: Int?
}

struct InstructorProfileUpdate: Encodable {
    let headline: String
    let bio: String
    let specialties: String
    let teachingYears: Int?
}

struct StripeStatus: Decodable {
    let connected: Bool?
    let onboarded: Bool?
}

struct Category: Decodable, Identifiable, Equatable {
    var id: String { slug }
    let name: String
    let slug: String
}

struct NewLesson: Encodable, Identifiable, Equatable {
    var id: String = UUID().uuidString
    var title = ""
    var type = "text" // text | video
    var order = 0
    var content: String?
    var videoUrl: String?
}

struct NewSection: Encodable, Identifiable, Equatable {
    var id: String = UUID().uuidString
    var title = ""
    var order = 0
    var lessons: [NewLesson] = []
}

struct NewCoursePayload: Encodable {
    let title: String
    let description: String
    let category: String
    let level: String
    let modality: String // virtual | evento
    /// Derivado de la modalidad: virtual → ondemand · evento → fechado
    let courseType: String
    let startDate: String
    let duration: String
    let price: Double
    let imageUrl: String?
    let sections: [NewSection]
    let isDraft: Bool
}

struct HostGame: Decodable, Identifiable {
    struct Count: Decodable { let participants: Int?; let questions: Int? }
    let id: String
    let title: String
    let code: String
    let status: String
    let _count: Count?
}

struct NewGameQuestion: Encodable, Identifiable, Equatable {
    var id = UUID()
    var question = ""
    var options = ["", "", "", ""]
    var correct = 0
    enum CodingKeys: String, CodingKey { case question, options, correct }
}

/// Estado del juego visto por el anfitrión (incluye código y total de preguntas).
struct HostGameState: Decodable {
    struct Game: Decodable {
        let id: String
        let status: String
        let currentQuestion: Int?
        let code: String?
        let questions: [GameQuestion]?
    }
    let game: Game
    let currentQ: GameQuestion?
    let participants: [GameParticipant]

    var ranked: [GameParticipant] { participants.sorted { $0.score > $1.score } }
}

// MARK: - Admin

struct AdminStats: Decodable {
    let totalUsers: Int
    let totalCourses: Int
    let publishedCourses: Int
    let draftCourses: Int
    let totalEnrollments: Int
    /// Valor de catálogo (precio × inscripciones), NO dinero cobrado.
    let estimatedRevenue: Double
}

struct AdminFinances: Decodable {
    let totalRevenue: Double?
    let totalPlatformFee: Double?
    let totalInstructorPayouts: Double?
    let thisMonthRevenue: Double?
}

struct AdminAnalytics: Decodable {
    struct Revenue: Decodable, Identifiable { var id: String { month }; let month: String; let amount: Double }
    struct Users: Decodable, Identifiable { var id: String { month }; let month: String; let users: Int }
    let revenueByMonth: [Revenue]?
    let usersByMonth: [Users]?
}

struct AdminReview: Decodable, Identifiable {
    struct User: Decodable { let name: String? }
    struct Course: Decodable { let title: String? }
    let id: String
    let rating: Int
    let comment: String?
    let approved: Bool
    let user: User?
    let course: Course?
}

struct AdminApplication: Decodable, Identifiable {
    struct User: Decodable { let id: String; let name: String?; let email: String? }
    let id: String
    let status: String
    let headline: String?
    let bio: String?
    let reason: String?
    let user: User?
}

struct AdminUser: Decodable, Identifiable {
    let id: String
    let name: String?
    let email: String?
    var role: String
}

struct AdminCoupon: Decodable, Identifiable {
    let id: String
    let code: String
    let discountPct: Int
    let maxUses: Int?
    let usedCount: Int?
    let active: Bool
}

struct AdminCategory: Decodable, Identifiable {
    struct Count: Decodable { let courses: Int? }
    let id: String
    let name: String
    let slug: String
    let _count: Count?
}

struct AdminKpi: Decodable, Identifiable {
    let id: String
    let name: String
    let unit: String?
    let targetValue: Double
    let currentValue: Double

    var percent: Int { targetValue > 0 ? min(100, Int((currentValue / targetValue * 100).rounded())) : 0 }
}

struct QuoteRequest: Decodable, Identifiable {
    let id: String
    let companyName: String
    let contactName: String
    let contactEmail: String
    let contactPhone: String?
    let companySize: String?
    let interests: String?
    let message: String?
    var status: String

    var statusLabel: String {
        ["new": "Nueva", "contacted": "Contactada", "converted": "Convertida", "closed": "Cerrada"][status] ?? status
    }
}

struct QuoteRequestPayload: Encodable {
    let companyName: String
    let contactName: String
    let contactEmail: String
    let contactPhone: String?
    let companySize: String?
    let interests: String?
    let message: String?
}

extension String {
    /// Slug como lo genera la app Expo al crear categorías.
    var slugified: String {
        let folded = folding(options: .diacriticInsensitive, locale: nil).lowercased()
        let parts = folded.split { !($0.isLetter || $0.isNumber) || !$0.isASCII }
        return parts.joined(separator: "-")
    }
}
